File: formcraft-plm/src/main/java/fr/formcraft/repo/audit/AuditService.java
package fr.formcraft.repo.audit;

import fr.formcraft.common.constants.RepoConsts;
import fr.formcraft.model.entity.AuditLog;
import fr.formcraft.repo.jpa.AuditLogRepository;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service for writing audit log entries.
 * Uses async persistence to avoid slowing down main operations.
 */
@Service
public class AuditService {

    private static final Log logger = LogFactory.getLog(AuditService.class);

    private static final String PRODUCT_TYPE = "Product";

    private final AuditLogRepository auditLogRepository;

    @Autowired
    public AuditService(AuditLogRepository auditLogRepository) {
        this.auditLogRepository = auditLogRepository;
    }

    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void logCreate(Long entityId, String details) {
        writeLog(entityId, PRODUCT_TYPE, RepoConsts.AUDIT_CREATE, details);
    }

    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void logUpdate(Long entityId, String details) {
        writeLog(entityId, PRODUCT_TYPE, RepoConsts.AUDIT_UPDATE, details);
    }

    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void logFormulation(Long productId, String chainId, String status) {
        writeLog(productId, PRODUCT_TYPE, RepoConsts.AUDIT_FORMULATE,
                "chainId=" + chainId + " status=" + status);
    }

    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void logWorkflowTransition(Long productId, String fromState, String toState) {
        writeLog(productId, PRODUCT_TYPE, RepoConsts.AUDIT_TRANSITION,
                fromState + " -> " + toState);
    }

    /** Generic audit entry for any module's entity (e.g. "ChangeRequest", "NonConformance", "Project"). */
    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void logAction(Long entityId, String entityType, String action, String details) {
        writeLog(entityId, entityType, action, details);
    }

    @Transactional(readOnly = true)
    public List<AuditLog> getProductHistory(Long productId) {
        return auditLogRepository.findByEntityIdAndEntityTypeOrderByPerformedAtDesc(productId, PRODUCT_TYPE);
    }

    @Transactional(readOnly = true)
    public List<AuditLog> getHistory(Long entityId, String entityType) {
        return auditLogRepository.findByEntityIdAndEntityTypeOrderByPerformedAtDesc(entityId, entityType);
    }

    @Transactional(readOnly = true)
    public List<AuditLog> getRecentActivity(int limit) {
        return auditLogRepository.findAllByOrderByPerformedAtDesc(
                org.springframework.data.domain.PageRequest.of(0, limit));
    }

    private void writeLog(Long entityId, String entityType, String action, String details) {
        try {
            AuditLog log = new AuditLog();
            log.setEntityId(entityId);
            log.setEntityType(entityType);
            log.setAction(action);
            log.setPerformedBy(getCurrentUser());
            log.setDetails(details);
            auditLogRepository.save(log);
        } catch (Exception e) {
            logger.error("Failed to write audit log for entityId=" + entityId + ": " + e.getMessage(), e);
        }
    }

    private String getCurrentUser() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        return (auth != null) ? auth.getName() : "system";
    }
}

File: app/services/audit_service.py
import logging
from threading import Thread

from sqlalchemy.orm import Session

from app.core.constants import (
    AUDIT_CREATE, AUDIT_DELETE, AUDIT_FORMULATE, AUDIT_QUALITY_CHECK,
    AUDIT_TRANSITION, AUDIT_UPDATE,
)
from app.core.database import SessionLocal
from app.models.audit_log import AuditLog
from app.repositories import audit_log_repository

logger = logging.getLogger(__name__)

PRODUCT_TYPE = "Product"


def _write_log_in_new_session(entity_id: int, entity_type: str, action: str, details: str | None, performed_by: str | None):
    """Write an audit log entry in a separate session (mirrors @Async + REQUIRES_NEW)."""
    try:
        db = SessionLocal()
        try:
            log = AuditLog()
            log.entity_id = entity_id
            log.entity_type = entity_type
            log.action = action
            log.performed_by = performed_by or "system"
            log.details = details
            audit_log_repository.save(db, log)
            db.commit()
        finally:
            db.close()
    except Exception as e:
        logger.error(f"Failed to write audit log for entityId={entity_id}: {e}")


def _fire_and_forget(entity_id: int, entity_type: str, action: str, details: str | None, performed_by: str | None = None):
    t = Thread(target=_write_log_in_new_session, args=(entity_id, entity_type, action, details, performed_by), daemon=True)
    t.start()


def log_create(entity_id: int, details: str, performed_by: str | None = None):
    _fire_and_forget(entity_id, PRODUCT_TYPE, AUDIT_CREATE, details, performed_by)


def log_update(entity_id: int, details: str, performed_by: str | None = None):
    _fire_and_forget(entity_id, PRODUCT_TYPE, AUDIT_UPDATE, details, performed_by)


def log_formulation(product_id: int, chain_id: str, status: str, performed_by: str | None = None):
    _fire_and_forget(product_id, PRODUCT_TYPE, AUDIT_FORMULATE, f"chainId={chain_id} status={status}", performed_by)


def log_workflow_transition(product_id: int, from_state: str, to_state: str, performed_by: str | None = None):
    _fire_and_forget(product_id, PRODUCT_TYPE, AUDIT_TRANSITION, f"{from_state} -> {to_state}", performed_by)


def log_action(entity_id: int, entity_type: str, action: str, details: str, performed_by: str | None = None):
    _fire_and_forget(entity_id, entity_type, action, details, performed_by)


def get_product_history(db: Session, product_id: int) -> list[AuditLog]:
    return audit_log_repository.find_by_entity(db, product_id, PRODUCT_TYPE)


def get_history(db: Session, entity_id: int, entity_type: str) -> list[AuditLog]:
    return audit_log_repository.find_by_entity(db, entity_id, entity_type)


def get_recent_activity(db: Session, limit: int = 20) -> list[AuditLog]:
    return audit_log_repository.find_all_ordered(db, offset=0, limit=limit)
